app/src/main/java/cloudcity/util/CellUtil.java

```diff
@@ -122,46 +89,121 @@ public static MeasurementsModel getRegisteredCellInformationUpdatedBySignalStren
+     * @return the {@link MeasurementsModel} obtained from information contained in the {@code currentCell}
+     */
+    public static @NonNull MeasurementsModel getMeasurementsModel(CellInformation currentCell, CellInformation signalStrength, CellInfoPrecedence precedence) {
+        MeasurementsModel measurements = new MeasurementsModel();
 
-        if (cellForUpdating instanceof LTEInformation) {
-            LTEInformation lteCell = (LTEInformation) cellForUpdating;
-            measurements.setRsrp(lteCell.getRsrp());
-            measurements.setRsrq(lteCell.getRsrq());
-            measurements.setRssnr(lteCell.getRssnr());
+        // Since all of these 'informations' are just numbers, they will never be NULL, however they
+        // will be CellInfo.UNAVAILABLE so we will use that as a NULL marker;
+        //
+        // The idea is simple, if both the 'currentCell' and 'signalStrength' contain a non-null
+        // value, take the one that takes precedence; otherwise take the non-null value.
+        // If both of them are null, well then, take the null that takes precedence since it changes nothing
+        //
+        // We are in a bad situation if currentCell and signalStrength can end up being of different types.
+
+        // New safety
+        if (currentCell instanceof NRInformation) {
+            NRInformation nrCell = (NRInformation) currentCell;
+            // Assume signalStrength is the same
+            NRInformation nrSignal = (NRInformation) signalStrength;
+
+            Integer validCsirsrp = getNonNullAndNonInvalidMember(nrCell.getCsirsrp(), nrSignal.getCsirsrp(), CellInfoPrecedence.SIGNAL_INFO);
+            Integer validCsirsrq = getNonNullAndNonInvalidMember(nrCell.getCsirsrq(), nrSignal.getCsirsrq(), CellInfoPrecedence.SIGNAL_INFO);
+            Integer validCsisinr = getNonNullAndNonInvalidMember(nrCell.getCsisinr(), nrSignal.getCsisinr(), CellInfoPrecedence.SIGNAL_INFO);
+            Integer validSsrsrp = getNonNullAndNonInvalidMember(nrCell.getSsrsrp(), nrSignal.getSsrsrp(), CellInfoPrecedence.SIGNAL_INFO);
+            Integer validSsrsrq = getNonNullAndNonInvalidMember(nrCell.getSsrsrq(), nrSignal.getSsrsrq(), CellInfoPrecedence.SIGNAL_INFO);
+            Integer validSssinr = getNonNullAndNonInvalidMember(nrCell.getSssinr(), nrSignal.getSssinr(), CellInfoPrecedence.SIGNAL_INFO);
+
+            measurements.setCsirsrp(validCsirsrp);
+            measurements.setCsirsrq(validCsirsrq);
+            measurements.setCsisinr(validCsisinr);
+            measurements.setSsrsrp(validSsrsrp);
+            measurements.setSsrsrq(validSsrsrq);
+            measurements.setSssinr(validSssinr);
+        } else if (currentCell instanceof LTEInformation) {
+            LTEInformation lteCell = (LTEInformation) currentCell;
+            // Likewise, assume signal strength is the same
+            LTEInformation lteSignal = (LTEInformation) signalStrength;
+
```

Comment:
thing is, if they are of different types, since `NRInformation` and `LTEInformation` share so little (and don't have the actual parameters we're looking for) I think it's in our best intention to just blindly assume they will be of the same type or crash if they're not. So that we know of that possibility.

I've thought about this and we simply don't have a valid solution or a good idea what to do if the currentCell and signalStrength are of different types. But we'll cross that bridge when we get to it, if we even get to it that is.